arrows/vxl/color_commonality_filter.cxx

```diff
@@ -0,0 +1,561 @@
+    }
+  }
+}
+
+// ----------------------------------------------------------------------------
+/// Create an output image indicating the relative commonality of each
+/// input pixel's color occuring in the entire input image. A lower
+/// value in the output image corresponds to that pixels value being
+/// less common in the entire input image.
+///
+/// Functions by first building a histogram of the input image, then,
+/// for each pixel, looking up the value in the histogram and scaling
+/// this value by a given factor.
+template < class InputType, class OutputType >
+void
+perform_filtering( vil_image_view< InputType > const& input,
+                   vil_image_view< OutputType >& output,
+                   color_commonality_filter_settings const& options )
+{
+  assert( std::numeric_limits< InputType >::is_integer );
+  assert( is_power_of_two( options.resolution_per_channel ) );
+
+  // Set output image size
+  output.set_size( input.ni(), input.nj() );
+
+  // If we are in grid mode, simply call this function recursively
+  if( options.grid_image )
+  {
+    color_commonality_filter_settings recursive_options = options;
+    recursive_options.grid_image = false;
+
+    // Formulate grided regions
+    unsigned ni = input.ni();
+    unsigned nj = input.nj();
+
+    for( unsigned j = 0; j < options.grid_resolution_height; ++j )
+    {
+      for( unsigned i = 0; i < options.grid_resolution_width; ++i )
+      {
+        // Top left point for region
+        int ti = ( i * ni ) / options.grid_resolution_width;
+        int tj = ( j * nj ) / options.grid_resolution_height;
+
+        // Bottom right
+        int bi = ( ( i + 1 ) * ni ) / options.grid_resolution_width;
+        int bj = ( ( j + 1 ) * nj ) / options.grid_resolution_height;
+
+        // Formulate rect region
+        vgl_box_2d< int > region{ ti, bi, tj, bj };
+
+        vil_image_view< InputType > region_data_ptr =
+          point_view_to_region( input, region );
+        vil_image_view< OutputType > output_data_ptr =
+          point_view_to_region( output, region );
+
+        // Process rect region independent of one another
+        perform_filtering( region_data_ptr, output_data_ptr,
+                           recursive_options );
+      }
+    }
+
+    // We are done, we already processed the entire image
+    return;
+  }
+
+  // Reset histogram (use internal or external version)
+  unsigned int hist_size = options.resolution_per_channel;
+
+  if( input.nplanes() == 3 )
+  {
+    hist_size = options.resolution_per_channel *
+                options.resolution_per_channel *
+                options.resolution_per_channel;
+  }
+
+  bool use_external_hist = ( options.histogram != NULL );
+
+  std::vector< unsigned >* histogram = options.histogram;
+
+  if( !use_external_hist )
+  {
+    histogram = new std::vector< unsigned >( hist_size );
+  }
+
+  histogram->resize( hist_size );
+
+  std::fill( histogram->begin(), histogram->end(), 0 );
```

Comment:
The paranoiac in me suggests:
```suggestion
  std::vector< unsigned >* histogram = options.histogram;
  std::unique_ptr< std::vector< unsigned > > local_histogram;

  if( !histogram )
  {
    local_histogram.reset( new std::vector< unsigned >{ hist_size } );
    histogram = local_histogram.get();
  }
  else
  {
    histogram->resize( hist_size );
    std::fill( histogram->begin(), histogram->end(), 0 );
  }
```
:slightly_smiling_face: 
(...and lose the later explicit delete, of course.)

Regardless, resizing and filling the vector if we just `new`d it is redundant.